### PR TITLE
initial POC commit for shared-config

### PR DESCRIPTION
**Pull request by rcoh**
## Motivation and Context
- for STS / SSO providers we need config loading split from the actual services
- Enable sharing a configuration between multiple services

This does not support the full scope planned in #655, specifically, there is no change to how connectors are handled or how the connector stack is built.

## Description
- Add `NowOrLater` utility to smithy-async
- Add `Config` to aws-types 
- Generate implementation of `From<&Config>` for AWS services
- Move _all_ credential providers into `aws-config`
- Change `AsyncProvideCredentials` to `ProvideCredentials`
- Change return type of `ProvideCredentials` to `future::ProvideCredentials`
- Add region provide from profile files

## Checklist
- [ ] I have updated the CHANGELOG

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/RawClientDecorator.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package software.amazon.smithy.rustsdk
+
+class RawClientDecorator
```
Comment (jdisanti) [inline aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/RawClientDecorator.kt:8]: What is this for?

File: rust-runtime/smithy-async/src/future/now_or_later.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    pub struct NowOrLater<T, F> {
+        #[pin]
+        inner: Inner<T, F>
+    }
+}
+
+pin_project! {
+    /// Utility Future
+    #[project = NowOrLaterProj]
+    pub enum Inner<T, F> {
```
Comment (jdisanti) [inline rust-runtime/smithy-async/src/future/now_or_later.rs:26]: Does this need to be public?

File: CHANGELOG.md [hunk 3]
```diff
@@ -123,7 +123,7 @@ v0.17 (July 15th 2021)
 **New this Week**
 
 - :tada: Add support for Autoscaling (#576, #582)
-- `AsyncProvideCredentials` now introduces an additional lifetime parameter, simplifying bridging it
+- `ProvideCredentials` now introduces an additional lifetime parameter, simplifying bridging it
```
[the remark this answers is not in the document]
Comment (rcoh) [reply CHANGELOG.md:126]: oh yeah whoops. global find and replace

File: aws/rust-runtime/aws-config/src/environment/credential.rs [hunk 4]
```diff
@@ -3,19 +3,37 @@
...
-use crate::Credentials;
+use aws_types::credential::provide_credentials::future;
+use aws_types::credential::{CredentialsError, ProvideCredentials};
 use aws_types::os_shim_internal::Env;
+use aws_types::{credential, Credentials};
 use std::env::VarError;
 
 /// Load Credentials from Environment Variables
-pub struct EnvironmentVariableCredentialsProvider {
+pub struct Provider {
```
[the remark this answers is not in the document]
Comment (rcoh) [reply aws/rust-runtime/aws-config/src/environment/credential.rs:13]: yeah I don't necessarily disagree. I was trying not to stutter but I found this nice rule of thumb:
> People will use module-qualified paths if they are sensible and pervasive in your documentation and usage examples. "Sensible" is a function of number of characters typed and how unique the name is.

> In my experience people are fine typing 2-3 character modules and mostly fine with 4 character modules. More than that is asking too much, meaning that items in a 5+ character module or library better have a name that can stand alone.

I think having the providers have full names is probably prudent.

File: rust-runtime/smithy-async/src/future/now_or_later.rs (hunk 2 above)
Comment (rcoh) [reply rust-runtime/smithy-async/src/future/now_or_later.rs:26]: nope good catch

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/RawClientDecorator.kt (hunk 1 above)
Comment (rcoh) [reply aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/RawClientDecorator.kt:8]: there's a refactoring in here to remove aws-hyper and instead have per-service defined middleware stacks (which is I think, what we want in the long term).

I think we can find a better name, but this is the "low level" client that exposes a `.call` method.

We could name it "tower service decorator" or something along those lines.

**Pull request closed without merging**